File: quickmedi-ai/services/interaction_service.py
"""
Interaction Service Module
Helper service for interaction-related operations using optimized data + AI
"""

from typing import Dict, List, Optional
import logging

from models.interaction_checker import InteractionChecker
from models.duplicate_checker import DuplicateChecker

logger = logging.getLogger(__name__)


class InteractionService:
    """
    Service for managing interaction checks
    Uses optimized data loader (190k+ interactions) for instant results
    """
    
    def __init__(self):
        self.interaction_checker = InteractionChecker()
        self.duplicate_checker = DuplicateChecker()
        logger.info("InteractionService initialized with optimized data")
    
    def check_drug_interactions(self, medicines: List[str]) -> Dict:
        """
        Check drug-drug interactions (instant with optimized data)
        
        Args:
            medicines: List of medicine names
            
        Returns:
            Interaction report
        """
        try:
            result = self.interaction_checker.check_drug_interactions(medicines)
            return result
        except Exception as e:
            logger.error(f"Error checking drug interactions: {str(e)}")
            raise
    
    def check_duplicates(self, medicines: List[str]) -> Dict:
        """
        Check for duplicate medicines (instant with optimized data)
        
        Args:
            medicines: List of medicine names
            
        Returns:
            Duplicate report
        """
        try:
            result = self.duplicate_checker.check_duplicates(medicines)
            return result
        except Exception as e:
            logger.error(f"Error checking duplicates: {str(e)}")
            raise
    
    def comprehensive_check(
        self, 
        medicines: List[str],
        user_conditions: Optional[List[str]] = None,
        user_allergies: Optional[List[str]] = None
    ) -> Dict:
        """
        Perform comprehensive safety check (optimized + AI insights)
        
        Args:
            medicines: List of medicine names
            user_conditions: User's medical conditions
            user_allergies: User's allergies
            
        Returns:
            Comprehensive safety report
        """
        try:
            results = {
                "medicines_checked": len(medicines),
                "safe": True,
                "warnings": [],
                "drug_interactions": {},
                "duplicates": {},
                "severity": "none"
            }
            
            # Check drug-drug interactions (instant)
            if len(medicines) >= 2:
                interactions = self.interaction_checker.check_drug_interactions(medicines)
                results["drug_interactions"] = interactions
                
                if interactions.get("has_interactions"):
                    results["safe"] = False
                    results["warnings"].append("Drug interactions detected")
                    
                    # Determine highest severity
                    for interaction in interactions.get("interactions", []):
                        severity = interaction.get("severity", "low")
                        if severity == "high":
                            results["severity"] = "high"
                        elif severity == "moderate" and results["severity"] != "high":
                            results["severity"] = "moderate"
                        elif results["severity"] == "none":
                            results["severity"] = "low"
            
            # Check duplicates (instant)
            if len(medicines) >= 2:
                duplicates = self.duplicate_checker.check_duplicates(medicines)
                results["duplicates"] = duplicates
                
                if duplicates.get("has_duplicates"):
                    results["safe"] = False
                    results["warnings"].append("Duplicate active ingredients detected")
            
            # Add user conditions/allergies warnings if provided
            if user_conditions:
                results["user_conditions"] = user_conditions
                results["warnings"].append("Check medicine compatibility with conditions")
            
            if user_allergies:
                results["user_allergies"] = user_allergies
                results["warnings"].append("Check for allergy risks")
            
            return results
            
        except Exception as e:
            logger.error(f"Error in comprehensive check: {str(e)}")
            raise

File: quickmedi-web/src/services/interaction.service.ts
/**
 * Interaction Service
 * Handles drug interaction checking using AI service
 */

import { apiClient } from '@/lib/api-client';
import { API_CONFIG } from '@/lib/api-config';
import type { ApiResponse } from '@/types/api-types';

// AI Interaction Types
interface DrugInteraction {
  medicine1: string;
  medicine2: string;
  severity: 'minor' | 'moderate' | 'severe';
  description: string;
  recommendation: string;
}

interface DuplicateCheck {
  medicine: string;
  is_duplicate: boolean;
  duplicate_of?: string;
  reason?: string;
}

interface ComprehensiveCheck {
  interactions: DrugInteraction[];
  duplicates: DuplicateCheck[];
  warnings: string[];
  safe_to_take: boolean;
}

export const interactionService = {
  /**
   * Check drug-drug interactions (382k+ database)
   */
  async checkDrugInteractions(medicines: string[]): Promise<ApiResponse<DrugInteraction[]>> {
    return apiClient.post(
      API_CONFIG.AI.INTERACTION.CHECK_DRUGS,
      { medicines },
      true
    );
  },

  /**
   * Check specific interaction between two medicines
   */
  async checkSpecificInteraction(medicine1: string, medicine2: string): Promise<ApiResponse<DrugInteraction>> {
    return apiClient.get(
      API_CONFIG.AI.INTERACTION.CHECK_SPECIFIC(medicine1, medicine2),
      true
    );
  },

  /**
   * Check for duplicate medicines
   */
  async checkDuplicates(medicines: string[]): Promise<ApiResponse<DuplicateCheck[]>> {
    return apiClient.post(
      API_CONFIG.AI.INTERACTION.CHECK_DUPLICATES,
      { medicines },
      true
    );
  },

  /**
   * Comprehensive check (interactions + duplicates + safety)
   */
  async comprehensiveCheck(
    medicines: string[],
    userConditions?: string[],
    userAllergies?: string[]
  ): Promise<ApiResponse<ComprehensiveCheck>> {
    return apiClient.post(
      API_CONFIG.AI.INTERACTION.COMPREHENSIVE,
      {
        medicines,
        user_conditions: userConditions,
        user_allergies: userAllergies,
      },
      true
    );
  },
};
